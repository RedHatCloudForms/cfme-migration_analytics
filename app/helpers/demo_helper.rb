module DemoHelper
  include_concern 'TextualSummary'
end
