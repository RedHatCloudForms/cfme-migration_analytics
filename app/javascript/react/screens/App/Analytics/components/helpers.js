export const readFirstFile = (fileHandles, onLoad) => {
  if (fileHandles && fileHandles.length > 0) {
    const reader = new FileReader();
    reader.onload = () => onLoad({ filename: fileHandles[0].name, body: reader.result });
    reader.readAsBinaryString(fileHandles[0]);
  }
};
