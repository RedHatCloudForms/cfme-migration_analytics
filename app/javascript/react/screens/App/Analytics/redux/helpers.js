export const calculateSummaryData = ({ vmSummaryReportResult, envSummaryReportResult }) => {
  const valuesByProviderId = {};

  vmSummaryReportResult.result_set.forEach(vm => {
    const {
      'ext_management_system.id': providerId,
      'ext_management_system.name': providerName,
      allocated_disk_storage: allocatedDiskSpace,
      mem_cpu: allocatedMemory,
      cpu_total_cores: numCpuCores
    } = vm;
    if (!valuesByProviderId[providerId]) {
      valuesByProviderId[providerId] = {
        id: providerId,
        name: providerName,
        numHypervisors: 0,
        numVms: 0,
        allocatedDiskSpace: 0,
        allocatedMemory: 0,
        numCpuCores: 0
      };
    }
    const provider = valuesByProviderId[providerId];
    provider.numVms += 1;
    provider.allocatedDiskSpace += allocatedDiskSpace;
    provider.allocatedMemory += allocatedMemory;
    provider.numCpuCores += numCpuCores;
  });

  envSummaryReportResult.result_set.forEach(hypervisor => {
    const providerId = hypervisor['ext_management_system.id'];
    valuesByProviderId[providerId].numHypervisors += 1;
  });

  const total = {
    numProviders: Object.keys(valuesByProviderId).length,
    numHypervisors: 0,
    numVms: 0,
    allocatedDiskSpace: 0,
    allocatedMemory: 0,
    numCpuCores: 0
  };

  Object.keys(valuesByProviderId).forEach(providerId => {
    const provider = valuesByProviderId[providerId];
    total.numHypervisors += provider.numHypervisors;
    total.numVms += provider.numVms;
    total.allocatedDiskSpace += provider.allocatedDiskSpace;
    total.allocatedMemory += provider.allocatedMemory;
    total.numCpuCores += provider.numCpuCores;
  });

  return {
    total,
    providers: Object.values(valuesByProviderId)
  };
};
