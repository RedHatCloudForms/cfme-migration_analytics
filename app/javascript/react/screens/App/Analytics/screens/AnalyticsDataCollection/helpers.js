const payloadPathRegex = /\/migration_analytics\/.+$/;

export const getPayloadUrl = payloadPath => {
  const match = payloadPath && payloadPath.match(payloadPathRegex);
  if (!match) return null;
  const { protocol, hostname, port } = window.location;
  const origin = `${protocol}//${hostname}${port ? `:${port}` : ''}`;
  return `${origin}${match[0]}`;
};
