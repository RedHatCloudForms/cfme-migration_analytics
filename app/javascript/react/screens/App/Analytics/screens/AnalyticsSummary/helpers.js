export const someProvidersExist = ({ providers }) => providers && providers.length > 0;

export const noRefreshErrorsExist = ({ providersWithRefreshErrors }) =>
  providersWithRefreshErrors && providersWithRefreshErrors.length === 0;

export const someProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length > 0;

export const noProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length === 0;

export const providersRefreshed = props =>
  someProvidersExist(props) && noRefreshErrorsExist(props) && noProvidersAwaitingRefresh(props);
