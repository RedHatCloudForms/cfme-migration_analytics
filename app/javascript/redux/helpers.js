import Immutable from 'seamless-immutable';
import URI from 'urijs';
import API from '../common/API';

export const functionLookupReducer = (initialState, actionHandlers) => (state = initialState, action) => {
  const handler = action && actionHandlers[action.type];
  return handler ? handler(state, action) : state;
};

export const getHandlersForFetchResourcesActions = (actionType, isFetchingKey, errorKey, resourcesKey) => ({
  initialState: Immutable({
    [isFetchingKey]: false,
    [errorKey]: null,
    [resourcesKey]: null
  }),
  actionHandlers: {
    [`${actionType}_PENDING`]: state => state.set(errorKey, null).set(isFetchingKey, true),
    [`${actionType}_FULFILLED`]: (state, action) =>
      state
        .set(errorKey, null)
        .set(isFetchingKey, false)
        .set(resourcesKey, action.payload.data.resources),
    [`${actionType}_REJECTED`]: (state, action) =>
      state.set(errorKey, action.payload.data.error).set(isFetchingKey, false)
  }
});

export const getHandlersForFetchActionsIndexedByHref = (actionType, fetchingHrefsKey, errorKey, payloadsByHrefKey) => ({
  initialState: Immutable({
    [fetchingHrefsKey]: [],
    [errorKey]: null,
    [payloadsByHrefKey]: {}
  }),
  actionHandlers: {
    [`${actionType}_PENDING`]: (state, action) =>
      state.set(errorKey, null).set(fetchingHrefsKey, [...state[fetchingHrefsKey], action.meta.href]),
    [`${actionType}_FULFILLED`]: (state, action) =>
      state
        .set(errorKey, null)
        .set(fetchingHrefsKey, state[fetchingHrefsKey].filter(href => href !== action.meta.href))
        .set(payloadsByHrefKey, { ...state[payloadsByHrefKey], [action.meta.href]: action.payload.data }),
    [`${actionType}_REJECTED`]: (state, action) =>
      state
        .set(errorKey, action.payload.data.error)
        .set(fetchingHrefsKey, state[fetchingHrefsKey].filter(href => href !== action.meta.href))
  }
});

export const formatApiFilterValues = filterValues =>
  Object.keys(filterValues).map(key => `${key}='${filterValues[key]}'`);

const resourceMatchesFilters = (resource, filterValues) =>
  Object.keys(filterValues).every(key => filterValues[key] === resource[key]);

export const filterResources = (resources, filterValues) =>
  resources && resources.filter(resource => resourceMatchesFilters(resource, filterValues));

export const findResource = (resources, filterValues) =>
  resources && resources.find(resource => resourceMatchesFilters(resource, filterValues));

export const fetchExpandedResourcesAction = (type, href, filterValues, attributes) => dispatch => {
  const uri = new URI(href);
  if (filterValues) uri.addSearch('filter[]', formatApiFilterValues(filterValues));
  if (attributes) uri.addSearch('attributes', attributes.join(','));
  uri.addSearch('expand', 'resources');
  return dispatch({
    type,
    payload: API.get(uri.toString())
  });
};

export const simpleActionWithProperties = (type, properties) => dispatch => dispatch({ type, ...properties });
