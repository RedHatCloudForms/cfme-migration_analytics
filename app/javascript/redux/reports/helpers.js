export const concatPreservingUniqueIds = (currentSet, newElements) => {
  const newIds = newElements.map(elem => elem.id);
  const currentSetWithConflictsRemoved = currentSet.filter(elem => !newIds.includes(elem.id));
  return [...currentSetWithConflictsRemoved, ...newElements];
};
