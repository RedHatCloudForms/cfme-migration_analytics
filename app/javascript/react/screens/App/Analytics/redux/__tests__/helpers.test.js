import { calculateSummaryData } from '../helpers';

describe('calculate summary data helper', () => {
  test('returns base case without errors when results are empty', () => {
    const results = {
      vmSummaryReportResult: { result_set: [] },
      envSummaryReportResult: { result_set: [] }
    };
    const summaryData = calculateSummaryData(results);
    expect(summaryData).toMatchSnapshot();
  });

  test('returns correct sums', () => {
    const results = {
      vmSummaryReportResult: {
        result_set: [
          {
            'ext_management_system.id': 1,
            'ext_management_system.name': 'provider 1',
            allocated_disk_storage: 2000,
            mem_cpu: 1000,
            cpu_total_cores: 1
          },
          {
            'ext_management_system.id': 1,
            'ext_management_system.name': 'provider 1',
            allocated_disk_storage: 2000,
            mem_cpu: 1000,
            cpu_total_cores: 1
          },
          {
            'ext_management_system.id': 2,
            'ext_management_system.name': 'provider 2',
            allocated_disk_storage: 2000,
            mem_cpu: 1000,
            cpu_total_cores: 1
          }
        ]
      },
      envSummaryReportResult: {
        result_set: [
          {
            'ext_management_system.id': 1
          },
          {
            'ext_management_system.id': 2
          },
          {
            'ext_management_system.id': 2
          }
        ]
      }
    };
    const summaryData = calculateSummaryData(results);
    const sumsFromProviders = { numProviders: summaryData.providers.length };
    summaryData.providers.forEach(provider => {
      Object.keys(provider).forEach(key => {
        if (key !== 'id' && key !== 'name') {
          if (!sumsFromProviders[key]) sumsFromProviders[key] = 0;
          sumsFromProviders[key] += provider[key];
        }
      });
    });
    expect(sumsFromProviders).toEqual(summaryData.total);
    expect(summaryData.total.numProviders).toEqual(
      new Set(results.vmSummaryReportResult.result_set.map(vm => vm['ext_management_system.id'])).size
    );
    expect(summaryData.total.numHypervisors).toEqual(results.envSummaryReportResult.result_set.length);
    expect(summaryData).toMatchSnapshot();
  });
});
