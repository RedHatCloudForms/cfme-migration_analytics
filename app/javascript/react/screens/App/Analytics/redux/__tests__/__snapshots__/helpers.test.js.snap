// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate summary data helper returns base case without errors when results are empty 1`] = `
Object {
  "providers": Array [],
  "total": Object {
    "allocatedDiskSpace": 0,
    "allocatedMemory": 0,
    "numCpuCores": 0,
    "numHypervisors": 0,
    "numProviders": 0,
    "numVms": 0,
  },
}
`;

exports[`calculate summary data helper returns correct sums 1`] = `
Object {
  "providers": Array [
    Object {
      "allocatedDiskSpace": 4000,
      "allocatedMemory": 2000,
      "id": 1,
      "name": "provider 1",
      "numCpuCores": 2,
      "numHypervisors": 1,
      "numVms": 2,
    },
    Object {
      "allocatedDiskSpace": 2000,
      "allocatedMemory": 1000,
      "id": 2,
      "name": "provider 2",
      "numCpuCores": 1,
      "numHypervisors": 2,
      "numVms": 1,
    },
  ],
  "total": Object {
    "allocatedDiskSpace": 6000,
    "allocatedMemory": 3000,
    "numCpuCores": 3,
    "numHypervisors": 3,
    "numProviders": 2,
    "numVms": 3,
  },
}
`;
